walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,151 @@
 
 package com.google.sps;
 
+import com.google.sps.TimeRange;
+
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
 
+/** Class responsible for scheduling a meeting between multiple people. */
 public final class FindMeetingQuery {
+  /**
+   * Return the times when a new meeting could happen in a day.
+   * 
+   * @param events  The collection of all known events in that day. Each event has:
+   *                a name, a time range and a collection of attendees.
+   * @param request The new meeting request, which has: 
+   *                a name, a duration in minutes and a collection of attendees.
+   *                The attendees are either mandatory or optional.
+   */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Set<Event> eventsSet = new HashSet<>(events);
+
+    // Get the people that should be attending this new meeting.
+    Set<String> mandatoryAttendees = new HashSet<>(request.getAttendees());
+    
+    // Get both mandatory and optional attendees.
+    Set<String> allAttendees = new HashSet<>(request.getAttendees());
+    allAttendees.addAll(request.getOptionalAttendees());
+
+    List<TimeRange> timeRangesForAllAttendees = 
+        solveQueryforAttendees(eventsSet, allAttendees, request.getDuration());
+    
+    // If one or more time slots exists so that both mandatory and optional attendees can attend, 
+    // return those time slots. 
+    if (!timeRangesForAllAttendees.isEmpty()) {
+      return timeRangesForAllAttendees;
+    }
+    
+    // Otherwise, return the time slots that fit just the mandatory attendees.
+    return solveQueryforAttendees(eventsSet, mandatoryAttendees, request.getDuration());
+  }
+
+  /**
+   * Return the times when a new meeting could happen in a day, so that all {@code attendees} can join.
+   *
+   * @param events    The set of all known events in that day.
+   * @param attendees The set of required attendees for the new meeting.
+   * @param duration  The time duration of the new meeting.
+   */
+  private static List<TimeRange> solveQueryforAttendees(Set<Event> events, Set<String> attendees, long duration) {
+    // Get the time ranges when at least one attendee is not free 
+    // (because for a particular time slot to work, all attendees must be free to attend the new meeting).
+    List<TimeRange> occupiedTimeRanges = events
```

Comment:
As commented below, it's better to sort in the "getNewMeetingTimeRanges" method. Otherwise, every other caller of this method would have to sort as well.

Reply:
Solved! :) thank you